File: python/s76.py
import collections


class Solution:
    def minWindow(self, s: str, t: str) -> str:
        need = collections.Counter(t)
        window = collections.defaultdict(int)
        valid = 0
        l, r = 0, 0
        a, b = -1, len(s)
        while r < len(s):
            c = s[r]
            r += 1
            if c in need:
                window[c] += 1
                if window[c] == need[c]:
                    valid += 1
            while l < r and valid == len(need):
                if  r - l < b - a:
                    a, b = l, r
                c = s[l]
                l += 1
                if c in need:
                    if window[c] == need[c]:
                        valid -= 1
                    window[c] -= 1
        return "" if a == -1 else s[a:b]


Solution().minWindow("ADOBECODEBANC", "ABC")

File: cpp-solve/leetcode/sliding_window/s76.cpp
#include "bits/stdc++.h"

using namespace std;

class Solution {
public:
    string minWindow(string s, string t) {
        map<char, int> window;
        map<char, int> need;
        for (char c:t) need[c]++;
        int valid = 0, l = 0, a = -1, b = s.size();
        for (int r = 0; r < s.size(); ++r) {
            char c = s[r];
            if (need.count(c)) {
                window[c]++;
                if (window[c] == need[c]) valid++;
            }
            while (valid == need.size()) {
                if (r - l < b - a) {
                    b = r;
                    a = l;
                }
                c = s[l]; // 记得改这里
                if (need.count(c)) {
                    if (window[c] == need[c]) valid--;
                    window[c]--;
                }
                l++;  // 记得更新
            }
        }
        return a == -1 ? "" : s.substr(a, (b - a) + 1);
    }
};

int main() {
    string ans = Solution().minWindow("ADOBECODEBANC",
                                      "ABC"
    );
    cout << ans;
}